### ISSUE TITLE
How do i do data augmentation before training with yolo

### ISSUE THREAD

**Issue by nil123532** [ultralytics/ultralytics#2701]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, Hello I want to do data transformations such as resize, random flips, color jitter and etc. 
I know in pytorch it can be done with the transforms package as follows:
# Performing Image Transformations. 
##Hints: Data Augmentation can be applied here. Have a look on RandomFlip, RandomRotation...
train_transform = transforms.Compose([
    #transforms.GaussianBlur(kernel_size=7, sigma=(0.5, 2)),
    transforms.Resize(112),
    #transforms.AutoAugment(),
    #transforms.RandomHorizontalFlip(),
    #transforms.RandomVerticalFlip(),
    #transforms.RandomRotation(15),
    #transforms.ColorJitter(brightness=0.5, contrast=0.7, saturation=0.4, hue=0.1),
    #
    transforms.ToTensor()
    #transforms.Normalize(([0.5126, 0.5122, 0.4562]), ([0.2791, 0.2659, 0.2897]))  
])

# Checking the dataset training size.
dataset = ImageFolder(data_dir, transform=train_transform)
print('Size of training dataset :', len(dataset))

But how can i do it for object detection in yolov8?

**Comment by glenn-jocher** [comment]
@nil123532 hello,

Thank you for reaching out. YOLOv8 uses Pytorch as the backend, so you can use the same Pytorch `transforms` package to perform image augmentations as you would for any other Pytorch model. 

One important thing to keep in mind is that YOLOv8 requires special care when performing data augmentations, since object detection datasets have both image and bbox annotations. You will need to ensure that the object annotations are updated correctly to match the augmented image. 

You can use the `imgaug` package, which is designed specifically for image augmentation for computer vision tasks. This package offers a range of transformations such as resize, horizontal/vertical flip, rotation, translation, and more. Additionally, the `imgaug` package has support for bounding boxes, which is a crucial aspect in object detection.

You can create a function that takes in an image and bounding box labels, and applies random augmentations to them. This way, you can generate multiple augmented versions of each image and corresponding bounding box labels, which can help prevent overfitting and improve the model's generalization performance during training.

I hope this helps. Let me know if you have any further questions!

**Comment by nil123532** [comment]
Thank you 👑. 
Just wondering if you could link me to a YouTube or Github repo link for this augmentation.

**Comment by glenn-jocher** [comment]
Hello @nil123532,

Thank you for reaching out to us. Unfortunately, we don't have any YouTube or Github repository links for data augmentation with YOLOv8 currently. However, data augmentation is a widely used technique in computer vision and there are many resources available online that can provide you with guidance on how to do this effectively. You can search for "Pytorch data augmentation for object detection" on Google or YouTube to find relevant tutorials and examples.

Additionally, the Pytorch `transforms` package can be used to perform data augmentation in YOLOv8 in the same way as for other Pytorch models. However, since YOLOv8 is an object detection model, you will need to make sure that any augmentations you apply to the images are consistent with the corresponding bounding box annotations. The `imgaug` package is one popular option for performing data augmentation with object detection models, and it provides a range of transformations specifically designed for this task.

I hope this helps! Let me know if you have any further questions.

**Comment by maichm** [comment]
It seems that data augmentation has been apply in v5augmentations.py when training.

**Comment by glenn-jocher** [comment]
@maichm hello,

Yes, data augmentation is applied during training in YOLOv8. The `v5augmentations.py` script contains the augmentation functions used for training. These include a variety of transformations, such as random resize, random flip, random crop, and random color jitter. 

Data augmentation is a crucial aspect of training object detection models such as YOLOv8. The main purpose of augmentations is to increase the variability of the training data and create more diverse examples for the model to learn from. This can help prevent overfitting and improve the model's generalization performance.

During training, each input image is randomly augmented using the functions defined in `v5augmentations.py`. The corresponding bounding box coordinates are updated accordingly to match the new augmented image. 

I hope this answers your question. Let me know if you have any further questions or concerns.

**Comment by rishiswethan** [comment]
@glenn-jocher It seems augmentations are applied by default. Is there any way I can configure which augmentations to use?

I found configs here, but not sure how where to apply it.
https://github.com/ultralytics/ultralytics/blob/main/docs/usage/cfg.md

**Comment by SaraDadjouy** [comment]
> @glenn-jocher It seems augmentations are applied by default. Is there any way I can configure which augmentations to use?
> 
> I found configs here, but not sure how where to apply it. https://github.com/ultralytics/ultralytics/blob/main/docs/usage/cfg.md

Hi. I have the same question. Did you find the answer?
